autoagora_agents/algorithm.py

```diff
@@ -0,0 +1,432 @@
+
+
+class PredeterminedAlgorithm(Algorithm):
+    """Change to a particular value at a given timestamp.
+
+    Attributes:
+        timestamps (list[int]): The timestamps at which to change the outputted value.
+            Must start with 0.
+        vals (list[np.ndarray]): The values outputted.
+    """
+
+    def __init__(
+        self, *, group: str, i: int, timestamps: list[int], vals: list[np.ndarray]
+    ) -> None:
+        super().__init__(group=group, i=i)
+        if timestamps[0] != 0:
+            raise ValueError("The first timestamp must be 0.")
+        self.timestamps = timestamps
+        self.vals = vals
+        self.ix = 0
+
+    def reset(self) -> None:
+        super().reset()
+        self.ix = 0
+
+    def __call__(
+        self,
+        *,
+        observation: np.ndarray,
+        action: np.ndarray,
+        reward: float,
+        done: bool,
+    ) -> np.ndarray:
+        if self.ix != len(self.timestamps) - 1:
+            if self.niterations >= self.timestamps[self.ix + 1]:
+                self.ix += 1
+
+        self.niterations += 1
+        return self.vals[self.ix]
+
+
+class BanditAlgorithm(Algorithm):
+    """Algorithms that have no observation other than the reward.
+
+    Keyword Arguments:
+        group (str): The group to which the algorithm belongs.
+        i (int): The id value of the object within the group.
+        bufferlength (int): The length of the buffer storing historical samples.
+        actiondistribution (dict): The config for the distribution representing the action.
+        optimizer (dict): The config for the optimizer.
+
+    Attributes:
+        actiondist (Distribution): The distribution modelling action-selection.
+        buffer (deque): The buffer storing historical samples.
+        optimizer (optim.Optimizer): A torch optimizer.
+    """
+
+    def __init__(
+        self,
+        *,
+        group: str,
+        i: int,
+        bufferlength: int,
+        actiondistribution: dict,
+        optimizer: dict,
+    ) -> None:
+        super().__init__(group=group, i=i)
+
+        self.actiondist = distributionfactory(**actiondistribution)
+        self.buffer = buffer.buffer(maxlength=bufferlength)
+        optimizer["params"] = self.actiondist.params
+        self.opt = optimizerfactory(**optimizer)
+
+    def reset(self):
+        super().reset()
+        self.actiondist.reset()
+        self.buffer.clear()
+
+    def __call__(
+        self,
+        *,
+        observation: np.ndarray,
+        action: np.ndarray,
+        reward: float,
+        done: bool,
+    ) -> np.ndarray:
+        act = np.array(self.actiondist.sample())
+        logprob = self.actiondist.logprob(torch.as_tensor(action))
+        self.buffer.append(
+            {
+                "reward": reward,
+                "action": action,
+                "logprob": logprob,
+            }
+        )
+        self.niterations += 1
+        return act
+
+    def logprob(self, actions):
```

Comment:
a tiny comment would be nice ;)

Reply:
46549629cb729eb7325e9552be6521fab8ef711c